torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -0,0 +1,202 @@
+        TORCH_INTERNAL_ASSERT(false); // not handled yet
+      } else {
+        if (auto ival = toIValue(node_->input(i))) {
+          replaceWithIValue(graph_->inputs().at(i), *ival);
+        }
+      }
+    }
+  }
+
+  c10::SymbolicShape run() {
+    // TODO: only run while the last iteration has made a change
+    for (size_t i = 0; i < 6; i++) {
+      substituteInputTensorProperties();
+      LowerSimpleTuples(graph_);
+      RemoveListMutation(graph_);
+      UnrollConstantLoops(graph_);
+      ConstantPropagation(graph_);
+      PeepholeOptimize(graph_);
+      ConstantPropagation(graph_);
+    }
+    ConstantPooling(graph_);
+    EliminateDeadCode(graph_);
+    return extractOutputShape();
+  }
+
+ private:
+  void substituteInputTensorProperties() {
+    for (auto index : tensor_indices) {
+      substituteTensorProperties(index);
+    }
+  }
+
+  void substituteTensorProperties(int64_t tensor_index) {
+    auto tensor_value = node_->inputs().at(tensor_index);
+    auto shape = tensor_value->type()->expect<TensorType>()->symbolic_sizes();
+    if (!shape.rank().has_value()) {
+      return;
+    }
+
+    for (const auto& use : graph_->inputs().at(tensor_index)->uses()) {
+      // TODO: either decompose composite ops like slice or add handling here
+      switch (use.user->kind()) {
+        case aten::len: {
+          size_t len = shape.rank().value();
+          replaceWithIValue(use.user->output(), static_cast<int64_t>(len));
+        } break;
+        case aten::__getitem__: {
+          auto index = constant_as<int64_t>(use.user->inputs().at(1));
+          if (index) {
+            auto norm_index = normIndex(*index, *shape.rank());
+            // TODO: HANDLE non-static value (symbolic shape)
+            if (norm_index && shape[*norm_index].is_static()) {
+              replaceWithIValue(
+                  use.user->output(), shape[*norm_index].static_size());
+            }
+          }
+        }
+      }
+    }
+  }
+
+  c10::SymbolicShape extractOutputShape() {
+    TORCH_INTERNAL_ASSERT(graph_->outputs().size() == 1);
+    auto output = graph_->outputs().at(0);
+    TORCH_INTERNAL_ASSERT(
+        output->type()->cast<ListType>() &&
+        output->type()->cast<ListType>()->getElementType()->cast<IntType>());
+    if (output->node()->kind() == prim::Constant) {
+      auto int_list = toIValue(output)->toIntVector();
+      return c10::SymbolicShape(int_list);
+    }
+    // If it is not a single list construct or constant, bail,
+    // otherwise we cannot analyze its output and it might be modified
+    if (output->node()->kind() != prim::ListConstruct ||
+        output->uses().size() != 1) {
+      return c10::SymbolicShape();
+    }
+    Node* list_construct = output->node();
+    std::vector<c10::optional<int64_t>> output_shape;
+    for (Value* input : list_construct->inputs()) {
+      output_shape.push_back(constant_as<int64_t>(input));
```

Comment:
Does `constant_as` throw in case the value is not constant? Should we add an explicit check for that?

Reply:
It returns c10::optional<int64_t>